src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.Analyzer.cs

```diff
@@ -0,0 +1,365 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.InvertIf
+{
+    internal sealed partial class CSharpInvertIfCodeRefactoringProvider
+    {
+        private sealed class Analyzer : Analyzer<IfStatementSyntax>
+        {
+            protected override string GetTitle()
+            {
+                return CSharpFeaturesResources.Invert_if;
+            }
+
+            protected override SyntaxNode GetRootWithInvertIfStatement(
+                Document document,
+                SemanticModel semanticModel,
+                IfStatementSyntax ifStatement,
+                InvertIfStyle invertIfStyle,
+                SyntaxNode subsequenceSingleExitPointOpt,
+                CancellationToken cancellationToken)
+            {
+                var generator = SyntaxGenerator.GetGenerator(document);
+                var syntaxFacts = CSharpSyntaxFactsService.Instance;
+
+                var ifNode = ifStatement;
+
+                var negatedCondition = (ExpressionSyntax)Negator.Negate(ifStatement.Condition, generator, syntaxFacts, semanticModel, cancellationToken);
+                var root = semanticModel.SyntaxTree.GetRoot(cancellationToken);
+
+                switch (invertIfStyle)
+                {
+                    case InvertIfStyle.Normal:
+                        {
+                            // For single line statement, we swap the TrailingTrivia to preserve the single line
+                            StatementSyntax newIfNodeStatement = null;
+                            ElseClauseSyntax newElseStatement = null;
+                            IfStatementSyntax oldIfStatement = ifStatement;
+
+                            var hasNewLineAfterClosingBrace = oldIfStatement.Statement.GetTrailingTrivia().Any(trivia => trivia.Kind() == SyntaxKind.EndOfLineTrivia);
+                            if (hasNewLineAfterClosingBrace)
+                            {
+                                newIfNodeStatement = oldIfStatement.Else.Statement.Kind() != SyntaxKind.Block
```

Comment:
part of me goes 'how do you know it has an else!', but then i assume that's implied by 'InvertIfStyle.Normal'.  But it's not clear.  That's why comments/examples go a long way.  It enables the readers/maintainers to know what is intentional vs. an accidental omission.

Reply:
needs explanation why it is appropriate to be creating a block if the 'if' when the else doesn't have a block. it's not obvious to me why that's correct, especially in the context of an 'invert' operation.

Reply:
(pending)